[watchtower/blob]: Blob of Justice encryption/decryption and v0 plaintext encoding

cfromknecht opened the pull request:
This PR introduces the encryption/decryption primtives for watchtower blobs, as well as the version 0 encoding scheme for the `blob.OfJustice` payload. The v0 encoding creates a constant size plaintext of 265 bytes, and allows a watchtower to sweep the to-local and (optionally) to-remote commitment outputs. Future versions can be made to support sweeping of HTLC outputs by using an alternate version. In the meantime, we choose a constant size encoding to not leak the number of outputs being swept.
```
blob of justice version 0 plaintext encoding:
   sweep address:               34 bytes
   revocation pubkey:           33 bytes
   local delay pubkey:          33 bytes
   csv delay:                    4 bytes
   commit local revocation sig: 64 bytes
   has p2wkh output:             1 byte
   p2wkh pubkey:                33 bytes, maybe blank
   p2wkh sig:                   64 bytes, maybe blank
```

The plaintext encoding is then enciphered using chacha20poly1305, using a 32-byte key and 12-byte nonce. The rationale for using chacha20poly1305 in this context is that the Lightning protocol already uses it for our brontide communication, meaning implementations can reuse existing primitives. In addition, chacha20poly1305 performs well on mobile devices, which is the primary audience for watchtower clients.

The ciphertext expansion of chacha20poly1305 adds 16-bytes to the length of the plaintext. Thus for the version 0 encoding, the encrypted blobs are a constant 281 bytes.
```
blob of justice version 0 ciphertext:
   enciphered plaintext: 265 bytes
   MAC:                   16 bytes
```

If someone wants to propose a better name than `blob.OfJustice`, I'm all ears. 🙉

jachiang commented:
Hi @cfromknecht , may I ask a clarifying question... The only sig in the blob is commit local delay sig, so a signature from the owner of the commitment TX? I thought the blob contained information for the watchtower to recreate the penalty transaction for revoked prior (counterparty) commitment TXs, so it would include revocation_sig etc. What am I missing? Still piecing together my understanding of watchtower ideas published so far, thanks in advance!

cfromknecht commented:
@jachiang yes you're correct, the signature should be a signature under the revocation key, for the remote party's commitment transcation. There's a typo in the PR body that makes it seem like it's a signature under the local delay key, thanks!

cfromknecht commented:
Will also move the typo fix into the godoc comments ;)

cfromknecht commented:
As per @halseth's request, removed the extra `HasP2WKH` flag from the serialization. A client now signals that the blob has a p2wkh output by ensuring that the first byte of the p2wkh pubkey field contains either 0x02 or 0x03. If the value contained is not one of these two, the decoded values will be discarded.

I've also added a method `HasP2WKHOutput()` that returns true if the p2wkh pubkey "looks" correct, using the same check applied during deserialization. This can be used after to decoding to determine if an attempt should be made to access the p2wkh pubkey/script/signature.

Plaintext v0 size is now 265 bytes, encrypted blob is then 281 bytes.
```
// blob version 0 plaintext encoding:
//    sweep address:               34 bytes
//    revocation pubkey:           33 bytes
//    local delay pubkey:          33 bytes
//    csv delay:                    4 bytes
//    commit local revocation sig: 64 bytes
//    p2wkh pubkey:                33 bytes, maybe blank
//    p2wkh sig:                   64 bytes, maybe blank
```

re: not having context of package prefix, I can't foresee a case in which it'd be used w/o it. as a standalone package, can assume that it will always be prefixed in usage. tbs, if the option is between `blob.JusticeKit` and `blob.OfJustice` i guess i'd lean towards former. makes the file names more logical

cfromknecht commented:
pushed up `blob.OfJustice --> blob.JusticeKit`

Roasbeef commented on watchtower/blob/justice_kit.go:
```diff
@@ -0,0 +1,406 @@
+	// encrypted with chacha20poly1305, which comes from a 16-byte MAC.
+	CiphertextExpansion = 16
+
+	// V0PlaintextSize is the plaintext size of a version 0 encoded blob.
```
Could probably copy over the comment in the body of the PR breaking down the side to this `godoc` string.

cfromknecht replied:
Done

Roasbeef commented on watchtower/blob/justice_kit.go:
```diff
@@ -0,0 +1,406 @@
+// and for a watchtower to later decrypt if action must be taken. The encoding
+// format is versioned to allow future extensions.
+type JusticeKit struct {
+	// SweepAddress is the p2wkh address of the output where the client's
```
Why 34 bytes?

cfromknecht replied:
Updated to 42 to accommodate maximally-sized witness programs

Roasbeef commented on watchtower/blob/justice_kit.go:
```diff
@@ -0,0 +1,406 @@
+// to-remote output.
+func (b *JusticeKit) CommitP2WKHScript() ([]byte, error) {
+	if btcec.IsCompressedPubKey(b.P2WKHPubKey[:]) {
+		return nil, nil
```
Looks like this should have a `!`, and also return an error in the failing case (so not a compressed key).

halseth commented on watchtower/blob/justice_kit.go:
```diff
@@ -0,0 +1,406 @@
+	)
+}
+
+// HasP2WKHOutput returns true if the blob contains a p2wkh pubkey and sig.
```
~~and sig~~? Comment a bit confusing since it is not checked in he code.
